infra/Pos/Slotting/Util.hs

```diff
@@ -137,40 +185,34 @@ onNewSlotImpl withLogging startImmediately action =
         -- REPORT:ERROR 'reportOrLogE' in 'onNewSlotImpl'
         reportOrLogE "Error occurred in 'onNewSlot' worker itself: " e
         delay =<< getNextEpochSlotDuration
-        onNewSlotImpl withLogging startImmediately action
+        onNewSlotImpl withLogging params action
 
 onNewSlotDo
-    :: OnNewSlot ctx m
-    => Bool -> Maybe SlotId -> Bool -> (SlotId -> m ()) -> m ()
-onNewSlotDo withLogging expectedSlotId startImmediately action = do
+    :: MonadOnNewSlot ctx m
+    => Bool -> Maybe SlotId -> OnNewSlotParams -> (SlotId -> m ()) -> m ()
+onNewSlotDo withLogging expectedSlotId onsp action = do
     curSlot <- waitUntilExpectedSlot
```

Comment:
Wait, so we will wait until current slot even if `onspStartImmediately onsp == True`? It seems that this call will block until new slot comes. Am I wrong?

Reply:
We won't wait until current slot if `expectedSlotId` is `Nothing`. However, we'll wait till the the end of recovery and it's you who introduced this logic 😉